pyomo/core/base/units_container.py

```diff
@@ -1285,33 +1343,222 @@ def check_units_equivalent(self, expr1, expr2):
+
+        """
+        src_pyomo_unit, src_pint_unit = self._get_units_tuple(src)
+        to_pyomo_unit, to_pint_unit = self._get_units_tuple(to_units)
+
+        # check if any units have offset
+        # CDL: This is no longer necessary since we don't allow
+        # offset units, but let's keep the code in case we change
+        # our mind about offset units
+        #  src_unit_container = pint.util.to_units_container(src_unit, self._pint_ureg)
+        # dest_unit_container = pint.util.to_units_container(dest_unit, self._pint_ureg)
+        # src_offset_units = [(u, e) for u, e in src_unit_container.items()
+        #                     if not self._pint_ureg._units[u].is_multiplicative]
+        # 
+        #  dest_offset_units = [(u, e) for u, e in dest_unit_container.items()
+        #                 if not self._pint_ureg._units[u].is_multiplicative]
+
+        # if len(src_offset_units) + len(dest_offset_units) != 0:
+        #     raise UnitsError('Offset unit detected in call to convert. Offset units are not supported at this time.')
+
+        # no offsets, we only need a factor to convert between the two
+        fac_b_src, base_units_src = self._pint_registry.get_base_units(src_pint_unit, check_nonmult=True)
+        fac_b_dest, base_units_dest = self._pint_registry.get_base_units(to_pint_unit, check_nonmult=True)
+
+        if base_units_src != base_units_dest:
+            raise UnitsError('Cannot convert {0:s} to {1:s}. Units are not compatible.'.format(src_unit, to_unit))
+
+        return fac_b_src/fac_b_dest*to_pyomo_unit/src_pyomo_unit*src
+
+    def convert_value(self, src, from_units=None, to_units=None):
+        """
+        This method performs explicit conversion of a numerical value (or
+        expression evaluated to a numerical value) from one unit to
+        another, and returns the new value.
+
+        If src is a native numerical type (e.g. float), then
+        from_units must be specified. If src is a Pyomo expression AND
+        from_units is None, then this code will retrieve the
+        from_units from the expression itself. Note that this method
+        returns a numerical value (not another Pyomo expression), so
+        any Pyomo expression passed in src will be evaluated.
+
+        If from_units is provided, but it does not agree with the units
+        in src, then an expection is raised.
+
+        Parameters
+        ----------
+        src : float or Pyomo expression
+           The source value that will be converted
+        from_units : None or Pyomo units expression
+           The units on the src. If None, then this method will try
+           to retrieve the units from the src as a Pyomo expression
+        to_units : Pyomo units expression
+           The desired target units for the new value
+
+        Returns
+        -------
+           float : The converted value
+
+        """
+        src_pyomo_unit, src_pint_unit = self._get_units_tuple(src)
```

Comment:
You get the src_units regardless of if you actually need them.  Should this only be done if either units checking is enabled or the user didn't provide from_units?

Reply:
I am going to rework these methods to simplify the api - it will remove this issue as well.